### PR TITLE
Script iframe intial load replaces about blank

### LINKED ISSUE #18564: need to navigate back three times to return back one page
STR:

1. Visit https://metajack.im
2. Click the link to the LCA talk
3. Attempt to go back to original page. It currently takes me 3 clicks of the back button. The first click does nothing, the second removes the image preview for the embed, and the third finally returns to the home page.

### PR DESCRIPTION
**Pull request by asajeffrey**
When navigating from the initial about:blank, do it with replacement enabled.

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #18564 
- [X] These changes do not require tests because the existing tests do the job

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/18587)

### CODE REVIEW

**Comment by asajeffrey** [comment]
I did a first pass of updating the test expectations, but it's not perfect.

**Comment by asajeffrey** [comment]
Only five failures!

**Comment by asajeffrey** [comment]
Hmm... fun with document loading:
```
$ RUST_LOG=script::document_loader ./mach test-wpt --debug-build --no-pause /fetch/api/request/multi-globals/url-parsing.html
...
 0:05.12 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:20715) Full command: /home/ajeffrey/github/asajeffrey/servo/target/debug/servo -w --hard-fail -u Servo/wptrunner -Z replace-surrogates -z http://web-platform.test:8000/fetch/api/request/multi-globals/url-parsing.html --user-stylesheet /home/ajeffrey/github/asajeffrey/servo/resources/ahem.css --certificate-path /tmp/tmp7uTOtg/cacert.pem
(pid:20715) "VMware, Inc."
 0:05.12 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:20715) "Gallium 0.4 on softpipe"
 0:05.12 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:20715) "3.3 (Core Profile) Mesa 17.2.0-devel"
 0:05.27 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:20715) "DEBUG:script::document_loader: Initial blocking load Some("http://web-platform.test:8000/fetch/api/request/multi-globals/url-parsing.html")."
 0:05.27 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:20715) "DEBUG:script::document_loader: Adding blocking load Script("http://web-platform.test:8000/resources/testharness.js") (1)."
 0:05.35 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:20715) "DEBUG:script::document_loader: Adding blocking load Script("http://web-platform.test:8000/resources/testharnessreport.js") (2)."
 0:05.35 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:20715) "DEBUG:script::document_loader: Removing blocking load Script("http://web-platform.test:8000/resources/testharness.js") (3)."
 0:05.36 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:20715) "DEBUG:script::document_loader: Adding blocking load Subframe("about:blank") (2)."
 0:05.36 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:20715) "DEBUG:script::document_loader: Initial blocking load Some("about:blank")."
 0:05.36 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:20715) "DEBUG:script::document_loader: Removing blocking load PageSource("about:blank") (1)."
 0:05.36 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:20715) "DEBUG:script::document_loader: Removing blocking load Subframe("about:blank") (3)."
 0:05.36 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:20715) "DEBUG:script::document_loader: Adding blocking load Subframe("http://web-platform.test:8000/fetch/api/request/multi-globals/incumbent/incumbent.html") (2)."
 0:05.37 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:20715) "DEBUG:script::document_loader: Removing blocking load PageSource("http://web-platform.test:8000/fetch/api/request/multi-globals/url-parsing.html") (3)."
 0:06.11 LOG: Thread-TestrunnerManager-2 WARNING u'runner_teardown' ()
 0:06.11 LOG: Thread-TestrunnerManager-2 INFO No more tests
 0:06.11 LOG: Thread-TestrunnerManager-3 WARNING u'runner_teardown' ()
 0:06.11 LOG: Thread-TestrunnerManager-3 INFO No more tests
 0:06.37 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:20715) "DEBUG:script::document_loader: Removing blocking load Script("http://web-platform.test:8000/resources/testharnessreport.js") (2)."
 0:07.06 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:20715) "DEBUG:script::document_loader: Initial blocking load Some("http://web-platform.test:8000/fetch/api/request/multi-globals/incumbent/incumbent.html")."
 0:07.07 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:20715) "DEBUG:script::document_loader: Adding blocking load Subframe("about:blank") (1)."
 0:07.07 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:20715) "DEBUG:script::document_loader: Initial blocking load Some("about:blank")."
 0:07.07 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:20715) "DEBUG:script::document_loader: Removing blocking load PageSource("about:blank") (1)."
 0:07.07 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:20715) "DEBUG:script::document_loader: Removing blocking load Subframe("about:blank") (2)."
 0:07.07 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:20715) "DEBUG:script::document_loader: Adding blocking load Subframe("http://web-platform.test:8000/fetch/api/request/multi-globals/current/current.html") (1)."
 0:07.08 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:20715) "DEBUG:script::document_loader: Removing blocking load PageSource("http://web-platform.test:8000/fetch/api/request/multi-globals/incumbent/incumbent.html") (2)."
 0:08.44 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:20715) "DEBUG:script::document_loader: Removing blocking load Subframe("http://web-platform.test:8000/fetch/api/request/multi-globals/incumbent/incumbent.html") (1)."
 0:09.13 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:20715) "DEBUG:script::document_loader: Initial blocking load Some("http://web-platform.test:8000/fetch/api/request/multi-globals/current/current.html")."
 0:09.14 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:20715) "DEBUG:script::document_loader: Removing blocking load PageSource("http://web-platform.test:8000/fetch/api/request/multi-globals/current/current.html") (1)."
 0:09.20 TEST_END: Thread-TestrunnerManager-1 Harness OK. Subtests passed 0/1. Unexpected 1
should parse the URL relative to the current settings object
------------------------------------------------------------
Expected PASS, got FAIL
promise_test: Unhandled rejection with value: object "TypeError: Url could not be parsed"
 0:09.20 LOG: Thread-TestrunnerManager-1 WARNING u'runner_teardown' ()
 0:09.20 LOG: Thread-TestrunnerManager-1 INFO No more tests
 0:10.21 LOG: Thread-TestrunnerManager-1 INFO Test runner process shut down
 0:10.21 LOG: Thread-TestrunnerManager-1 WARNING More tests found, but runner process died, restarting
 0:10.21 LOG: Thread-TestrunnerManager-1 INFO No more tests
 0:10.21 LOG: MainThread INFO Got 1 unexpected results
 0:10.21 SUITE_END: MainThread 
Summary
=======

Ran 2 tests (1 parents, 1 subtests)
Expected results: 1
Unexpected results: 1 (FAIL: 1)

Unexpected Results
==================

/fetch/api/request/multi-globals/url-parsing.html
-------------------------------------------------
FAIL should parse the URL relative to the current settings object
 0:10.25 LOG: MainThread INFO Closing logging queue
 0:10.25 LOG: MainThread INFO queue closed
```

The document loader for `incumbent.html` completes before the loader for `current.html` starts. Not good.

**Comment by asajeffrey** [comment]
IRC conversation with @cbrewster: http://logs.glob.uno/?c=mozilla%23servo&s=21+Sep+2017&e=21+Sep+2017#c756805

TL;DR: https://github.com/servo/servo/blob/581f0bf09a8ec35dadd5de207777f79138255e1a/components/script/dom/htmliframeelement.rs#L307-L312

Dubious, as an experiment I'm removing it, but @cbrewster reports that @jdm might not be happy about this.

@bors-servo: try

**Comment by asajeffrey** [comment]
We were sending a mozbrowserlocationchange event when changing the session history: https://github.com/servo/servo/blob/29517d553e6c2fddc7e3cf0c0abef58c6d0f34aa/components/constellation/constellation.rs#L2507

even though we were traversing the history in the case of a navigation with replacement enabled: https://github.com/servo/servo/blob/29517d553e6c2fddc7e3cf0c0abef58c6d0f34aa/components/constellation/constellation.rs#L2467

which already triggered a location change event: https://github.com/servo/servo/blob/29517d553e6c2fddc7e3cf0c0abef58c6d0f34aa/components/constellation/constellation.rs#L2370

This was triggering 2 mozbrowserlocationchange events.

File: components/constellation/constellation.rs [hunk 1]
```diff
@@ -2504,7 +2504,10 @@ impl<Message, LTF, STF> Constellation<Message, LTF, STF>
 
         // If the navigation is for a top-level browsing context, inform mozbrowser
         if change.browsing_context_id == change.top_level_browsing_context_id {
-            self.trigger_mozbrowserlocationchange(change.top_level_browsing_context_id);
+            // If this is a replacement then we've already triggered the event in traverse_to_entry
```
Comment (cbrewster) [inline components/constellation/constellation.rs]: Should we just return early when calling `traverse_to_entry`?

**Comment by asajeffrey** [comment]
On thinking about it, it seems like this PR is just a band-aid, and really we need to have a good look at our whole iframe loading story, with an eye to trying to align it better to the spec.
